core/types/calls.cc

```diff
@@ -1363,6 +1363,26 @@ class Magic_buildArray : public IntrinsicMethod {
     }
 } Magic_buildArray;
 
+class Magic_buildRange : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        if (args.args.empty()) {
+            res.returnType = Types::rangeOfUntyped();
+            return;
+        }
+
+        auto type_member = Types::dropLiteral(args.args[0]->type);
+
+        if (type_member->isNilClass()) {
+            type_member = Types::untypedUntracked();
+        } else if (Types::isSubType(gs, Types::nilClass(), type_member)) {
+            type_member = Types::dropNil(gs, type_member);
+        }
+
+        res.returnType = Types::rangeOf(gs, type_member);
```

Comment:
After this lands, will your other PR update to check the type of the second arg in the event that the first is nil?

Reply:
Yes, as this won't even parse at this stage.

I added an intermediary version to check the type of the second argument and return `T::Range[T.untyped]` if it's different from the first one.